Conversion of Preferences/TableColumnsTab to mvvm

calixtus opened the pull request:
follow up to #5033 , concerns #4927 #3416
fixes #5164 #4752 #3354 #4459

This is going to be a huge one. I already put some work into, but I need to take a little break, So I present an early WIP of the reworked TableColumsTab in the Preferences. There are some issues in sight, eg. the checkModel is sime kind of special.

I realized, there is maybe something wrong with the serializeSpecialFields-Option. I was not able to find any use for this pref in the whole code.

Im trying not to break anything, so besides the conversion I will not change the logic of what is controlled here. I'll leave that to another person, after im done here. My goal is to include all the specialFields and the FileFields in the ~CheckListView~, where you can sort, activate, deactivate, add custom fields (so called 'UnknownField') and remove them as you like.

Feel free to comment on it, give suggestions and ideas.

Update: We did some conceptual changes. See below.

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [x] Screenshots added in PR description (for bigger UI changes)
- [x] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

Siedlerchr commented:
I generally like the checkedListView approach, but I am not sure if we need all the file type columns. Maybe one generic (not sure if this is possible with our architecture)

Second, as we can reorder the columns in the maintable, it would be even better to add a righ click option there to add new columns instead of the preferences.
What do you think @JabRef/developers  @MartinKarim

calixtus commented:
I agree, adding and removing columns by a ContextMenu is probably more intuitive, and I must confess, I do not dislike this idea. One could also change the title or the type of the column by doubleclicking it or by dropdownMenu inside the ColumnTitle or something like that...
On the other hand, it can become confusing, of the list is options is growing to large and this would also break up the unity of the preferenceSettings. So it must really be worth it. Ideas? Suggestions? Opinions?

I put the File Columns in the CheckListView, to simplify the layout of the tab, as I found it very confusing with multiple listviews etc., because basically every single option (besides the radiobuttons) is creating a column in the maintable. So I think, the PrefsTab should reflect that. I wanted to change the checkboxes to simply enable the SpecialFields and FileFields as possible options in the checkListView.

About the ExtraFileColums: i was wondering myself, why there are distinct columns for every single ExternalFileType, as there is already a column, that does what you described, called the fileColumn. If only one file is attached, it displays a symbol of that type, if multiple files are attached, a generic symbol. By rightclicking that symbol, you can access the attached files. So I was not sure, if those multiple ExtraFileColumns are there as an artifact of an older JabRefVersion, or an enhancement, I did not yet understand, or just some silly idea of earlier days. 😉 This ExtraFileColumsLogic can probably be removed, as they are superceeded by the single fileColumn, can't it? Opinions?

Siedlerchr commented:
>. I wanted to change the checkboxes to simply enable the SpecialFields and FileFields as possible checkable options in the checkListView.

Yep this is also an idea I had an mind. 

>This ExtraFileColumsLogic can probably be removed, as they are superceeded by the single fileColumn, can't it? Opinions?

I would remove it and just keep the generic file column

Renaming columns is a bit problematic I think, as they are bound to the field constants. 

For the moment I would really say go with the single checked list view. We can then further discuss how to proceed with the maintable option contextmenu.

calixtus commented:
This is driving me mad. I'm debugging this for hours.
1. After storing the correct new order of the tablecolumns in the preferences, AND updating the maintable, the maintable uses for half a second the new columnorder, but then refreshes and uses the old one.
2. Moving a column in the prefs down leaves unchecked items in the checkListView. Not sure if this is my error, or a bug in controlsFX.

@JabRef/developers Suggestions? Ideas?

If it's ok for you, im going to delete the extraFileColumns as @Siedlerchr suggested...

calixtus commented:
Ad 1: I managed to locate the problem to the ChangeListener in PersistentVisualState of the MainTable. It somehow rewrites the preferences everytime a value is changed, so every change made in the preferences are rewritten, as soon as the columns of the MainTable are changed in the same session. My question is: As this is not directly a problem of the preferencesTab, but of the mainTable, should I go deep into it, fixing it (which would require some larger refactoring of the MainTable and the PersistentVisualState, or should I leave it, create in issue and concentrate on the other PreferencesTabs?
Ad 2: The checks are set correctly in the CheckViewModel, but they somehow won't be displayed in the ListView.  Moving items upwards on the other hand works correctly. This is very confusing to me.

calixtus commented:
Another solution would be to remove the up/down buttons in the preferences dialog and let the mainTable itself handle resizing and rearranging the columns (which it already does fine). This could be an early, workable and mergable solution. Someone could still add this functionality later, and I could move on to the next tab...
Opinions

calixtus commented:
I thought a while about the CheckListView. ControlsFX is somewhat nice, but it acutally causes more problems, than it solves. I'm going to change that to a tableView similiar to the XmpPrefsTab. But not today. 😄

tobiasdiez commented:
Sorry for the late answer, was hiking the last few weeks. First of all thanks a lot for the work you put into the conversation of the preference tabs. Much appreciated.

The table columns tab is a huge mess and your changes are a huge improvement. I think there are a few things that can still be improved:
- There is no conceptual difference between normal fields (e.g. author) and the special columns (e.g. doi/url, file, arxiv). In the end, both kinds are represented as columns - just that the display is slightly different for some columns. Thus, I would prefer a gui solution that treats all kinds of columns on the same level. Proposal: remove of all the special checkboxes on the right and integrate them in the main list view. Maybe it is a good idea to highlight the special fields when adding a new column (e.g. by having a drop-down control that first lists the special fields and then the ordinary ones).

- I think the additional checkbox adds a bit too much complexity to the interface. What does it mean to have a field/column in the list view which is unchecked? Is it shown in the main interface or not? In a special way? I would only display the fields/columns that are really shown and list the other options only if the user adds a new column.

- Having the functionality to move fields up and down is relatively important I would say (because for some users drag & drop of columns isn't that natural). You can also implement them as buttons in the row of the listview (similar to how it is done in the Manage journal abbreviations or key bindings dialog).

- Right now the functionality of the special fields is tightly connected to display of these fields as columns. In my opinion this should be decoupled and the preference settings connected to the special fields should be moved to a different tab.

- Concerning the bug with the PersistentVisualState, it is ok if you only create a new issue for it at the moment. (Although fixing it would be better of course.)

calixtus commented:
No problem, there is no hurry - summertime is vacation time.
Thanks for the suggestions. Im going to abandon the CheckListView, as it is problematic to handle and obviously a bit buggy, and the CheckBoxes. Instead I'm going to take a shot at the solution the journal abbreviations, with an actions-column.
About the special fields: I see, the right place for the contents of a special field to be shown, would be the entry-editor. Enabling them should therefore move either to the entryEditorTab oder the generalTab.

calixtus commented:
Issue seems to exist already: #5164
I think I'm gonna work on it.

calixtus commented:
ok, I changed the CheckListView to a TableView, but there is still work to do, e.g. including file, arxiv etc. in the combo, moving specialfields, fixing bugs.
But here is something to of my progress to see:

<img width="510" alt="tablecolumns" src="https://user-images.githubusercontent.com/50491877/63718156-37c05f00-c84a-11e9-954b-992697f075b5.png">

calixtus commented:
Ok, I fixed the bugs. turned out it wasn't the PersistentVisualState, but the BasePanel, that got refreshed with old preferences. There is one funny bug, that probably hasn't got to do something with my changes: In a very long list in the combobox, the grey background-hover is shown on the very opposite end of the visible list.

What I thought of is this: The tableColumnsTab is now converted and I dont want this PR to become to complex and large. 
I like to proposals @tobiasdiez made, but I would like to finish converting the other tabs first, so the preferences are basically ready for a beta-release. The other tabs should be quite easy to convert. 
After that im going to work on the MainTableColumnFactory. My Idea would be then, to create internal two pseudo internal fields for the files and the uris, so one could add or remove them in the table column preferences.

If this is fine with you, this would be ready to review.

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/TableColumnsTabView.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    private void setupIconButtons() {
+        ActionFactory actionFactory = new ActionFactory(Globals.getKeyPrefs());
```
You should get the keyPrefs also from the preferences Object

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/TableColumnsTabViewModel.java:
```diff
@@ -0,0 +1,318 @@
+
+    public BooleanProperty extraFileColumnsEnabledProperty() { return this.extraFileColumnsEnabledProperty; }
+
+    public class ExtraFileField implements Field {
```
I would actually move this class to a new file and put it next  to the other Fields.

calixtus replied:
I am somewhat unsure about this. As this is just a wrapper class for the TableRowModel in the columnsList to draw the ExtraFileFieldColumns, I dont think it belongs into the model of Jabref. We already talked about removing these ExtraFileColumns completly, but after the discussion in #5244 i am very careful about removing any functionality...

calixtus replied:
I created a util-class to store the ExtraFilePseudoField-class and a helper method which is going to be used also in PR #5265 . I hope this is a usable compromise.

tobiasdiez commented on src/main/java/org/jabref/gui/preferences/TableColumnsTabView.java:
```diff
@@ -0,0 +1,145 @@
+                        ((TableColumnsTabViewModel) viewModel).removeColumn(columnsList.getFocusModel().getFocusedItem()))
+                .install(actionsColumn);
+
+        ((TableColumnsTabViewModel) viewModel).selectedColumnModelProperty().setValue(columnsList.getSelectionModel());
```
For the next PR: you can get rid off these castings as follows:
- Add a generic type argument: `AbstractPreferenceTabView<T extends PreferenceTabViewModel>`
- Change the variable to `protected T viewModel;`
- Change declaration to `TableColumnsTabView extends AbstractPreferenceTabView<TableColumnsTabViewModel>`

tobiasdiez commented:
The code looks good to me and the UI is definitely a huge improvement already. Thus I'll merge now. It would be nice however if the changes proposed in https://github.com/JabRef/jabref/pull/5185#issuecomment-524642999 are not forgotten ;-)

Moreover, https://github.com/JabRef/jabref/pull/5185#discussion_r322013878 should give you few more minutes away from thesis ;-)
